### ISSUE TITLE
Add compatibility with fzf-tab and other plugins that reassign the tab key and/or redefine `_main_complete`/`compadd`

### ISSUE THREAD

**Issue by marlonrichert** [marlonrichert/zsh-autocomplete#501]
### Discussed in https://github.com/marlonrichert/zsh-autocomplete/discussions/500

<div type='discussions-op-text'>

<sup>Originally posted by **fujianyawa** October 24, 2022</sup>
Hello, thanks for the amazing plugin!
But how can I make  [fzf-tab](https://github.com/Aloxaf/fzf-tab) works alongside zsh-autocomplete?

[...]

Uncommenting this
`source $ZSH/custom/plugins/zsh-autocomplete/zsh-autocomplete.plugin.zsh` and  `plugins (#zsh-autocomplete ...)`
Makes the fzf-tab to work, but I want to use zsh-autocomplete too.

And also the author of the fzf-tab suggest [in this thread](https://github.com/Aloxaf/fzf-tab/issues/198#issuecomment-1288358158) to either choose which one of plugins to use, Though It's nice to use both of the plugins.

</div>

**Comment by p0da** [comment]
I am also interested in this but I didn't quite understand the explanation at https://github.com/marlonrichert/zsh-autocomplete/discussions/500#discussioncomment-3952103 describing how the requested feature would work. Would you mind clarifying?

**Comment by marlonrichert** [comment]
What @fujianyawa wants is to see autocompletions while typing, but use Fzf Tab when pressing Tab.

**Comment by fujianyawa** [comment]
> What @fujianyawa wants is to see autocompletions while typing, but use Fzf Tab when pressing Tab.

Uhm yes, basically `fzf-tab` for interactive fuzzy search completion when tabbing and `zsh-autocomplete` for ahead completion when typing


> I am also interested in this but I didn't quite understand the explanation at [#500 (reply in thread)](https://github.com/marlonrichert/zsh-autocomplete/discussions/500#discussioncomment-3952103) describing how the requested feature would work. Would you mind clarifying?

Oops my bad, My english skills are terrible, and I've worded it out poorly.

**Comment by kinoute** [comment]
I am also interested in this, that would be great to combine these two plugins

**Comment by marlonrichert** [comment]
What needs to be done for this:
* [ ] #418 
* [ ] Make a copy of Zsh's default `_main_complete` (and not whatever function currently has that name), so that we can call it and don't have to worry about other plugins overriding it.
* [ ] Keep our own wrapper for `_main_complete`, but don't replace any existing `_main_complete` with it, so we don't interfere with other plugins.
* [ ] Change our widgets to not call `_main_complete` but our wrapper instead, which in turns calls the copy we made of Zsh's own `_main_complete`.

Pull requests welcome! 🙂

**Comment by marlonrichert** [comment]
@kinoute @dhorkin @p0da @fujianyawa @dhorkin Can you check out the [`feature` branch](https://github.com/marlonrichert/zsh-autocomplete/tree/feature) and see if it works works out of the box there?

**Comment by marlonrichert** [comment]
@dhorkin No, the problem is that Fzf Tab doesn't know how to parse the [prompt escape sequences](https://zsh.sourceforge.io/Doc/Release/Prompt-Expansion.html) that are allowed in the completion system's [`format` style](https://zsh.sourceforge.io/Doc/Release/Completion-System.html#index-format_002c-completion-style). You should file a bug with Fzf Tab about this.

**Comment by dhorkin** [comment]
> @kinoute @dhorkin @p0da @fujianyawa @dhorkin Can you check out the [`feature` branch](https://github.com/marlonrichert/zsh-autocomplete/tree/feature) and see if the Tab key works out of the box there?

no joy, unfortunately :( it looks like the correct key is now being bound (output from `bindkey '^I'` is the same with or without `zsh-autocomplete`), but it isn't working. The compadd workaround is still needed:
```
my-fzf-tab() {
  functions[compadd]=$functions[-ftb-compadd]
  zle fzf-tab-complete
}
zle -N my-fzf-tab
bindkey "^I" my-fzf-tab
```

**Comment by dhorkin** [comment]
> @kinoute @dhorkin @p0da @fujianyawa @dhorkin Can you check out the [`feature` branch](https://github.com/marlonrichert/zsh-autocomplete/tree/feature) and see if the Tab key works out of the box there?

no joy, unfortunately 😔 it looks like the correct key is now being bound (output from `bindkey '^I'` is the same with or without `zsh-autocomplete`), but it isn't working. The compadd workaround is still needed:
```
my-fzf-tab() {
  functions[compadd]=$functions[-ftb-compadd]
  zle fzf-tab-complete
}
zle -N my-fzf-tab
bindkey "^I" my-fzf-tab
```

As a side note, I'm also using fz. My Antidote .zsh_plugins.txt:
```
skywind3000/z.lua
changyuheng/fz kind:defer
Aloxaf/fzf-tab
marlonrichert/zsh-autocomplete branch:feature
```

**Comment by marlonrichert** [comment]
> The `compadd` workaround is still needed:

Ah, I focused on `_main_complete` only and forgot about `compadd`. I can fix it in the same way, though.

**Comment by marlonrichert** [comment]
Actually, I think I found an even better way: I could do the patching by means of a subprocess, so that the state of the main shell is not affected.

**Comment by marlonrichert** [comment]
Should be fully working now. Let me know if it's not.

**Comment by yrammos** [comment]
@marlonrichert @dyhork I just installed `fzf-tab`. Is any setup necessary on the side of `zsh-autocomplete` to make it all work? Pressing tab currently opens `fzf-tab` but I don't see autocompletions while typing. Many thanks.

**Comment by marlonrichert** [comment]
> Pressing tab currently opens `fzf-tab` but I don't see autocompletions while typing.

That's strange. Please open a separate issue for this.

**Comment by yrammos** [comment]
@marlonrichert thanks for the clarification.

> That's strange. Please open a separate issue for this.

False alarm, my apologies — it now works as expected.

**Comment by kareefardi** [comment]
@yrammos mind telling my what was your issue. I am trying to setup up both plugins but I am not getting suggestion previews from zsh-autocomplete.

**Comment by kareefardi** [comment]
@marlonrichert looks like there is no issue. I somehow had a bad submodule and it wasn't syncing with master properly.

**Comment by marlonrichert** [comment]
Good to hear that it works! 🥳 

PS:
@fujianyawa @aaronkollasch @alexwforsythe @p0da @JTruj1ll0923 @cands @kinoute @dyhork @yrammos @kareefardi
If you enjoy using my software, please consider [sponsoring me](https://github.com/sponsors/marlonrichert). 💝
